prerelease_script: Problem with looking up valid OS codes

Apparantly there is a problem in looking up valid OS codes when creating pre-release jobs for distros other than local build machine's distro. My local build machine runs 64 bit _Ubuntu Trusty_ and has been setup to run prerelease jobs based on [the documentation on ROS wiki](http://wiki.ros.org/regression_tests#How_do_I_setup_my_system_to_run_a_prerelease.3F).

The prerelease command works fine when target distro is set to _Trusty_. However, when the target distro is set to _Vivid_ or _Utopic_ the generated command fails with the following error:

```
$ generate_prerelease_script.py \
>   https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/master/index.yaml \
>   jade default ubuntu utopic amd64 \
>   ardrone_autonomy \
>   --level 1 \
>   --output-dir .
Fetching buildfarm configuration...
Fetching rosdistro cache...
Evaluating job templates...
Traceback (most recent call last):
  File "/usr/bin/generate_prerelease_script.py", line 263, in <module>
    sys.exit(main())
  File "/usr/bin/generate_prerelease_script.py", line 157, in main
    source_repository=source_repository)
  File "/usr/lib/python2.7/dist-packages/ros_buildfarm/devel_job.py", line 237, in configure_devel_job
    ', '.join(sorted(build_file.targets[os_name].keys())))
ros_buildfarm.common.JobValidationError
```

Is there any way to setup a local and [minimal development environment to debug this issue? If so, I can try to track down the bug and fix it.

I quickly tried the second approach by forking the config repo and modifying `index.yaml` to re-use `release-build` config. It fails in `source_build_file.py` with this error: `AssertionError: Expected file type is 'source-build', not 'release-build'` (The types are indeed different)

--- a/ros_buildfarm/devel_job.py
+++ b/ros_buildfarm/devel_job.py
@@ -183,7 +183,8 @@ def configure_devel_job(
         jenkins=None, views=None,
         is_disabled=False,
         groovy_script=None,
-        source_repository=None):
+        source_repository=None,
+        build_targets=None):
     """
     Configure a single Jenkins devel job.
 
@@ -198,6 +199,9 @@ def configure_devel_job(
     if build_file is None:
         build_files = get_source_build_files(config, rosdistro_name)
         build_file = build_files[source_build_name]
+    # Overwrite build_file.targets if build_targets is specified
+    if build_targets is not None:
+        build_file.targets = build_targets
 
     if index is None:
         index = get_index(config.rosdistro_index_url)

--- a/scripts/prerelease/generate_prerelease_script.py
+++ b/scripts/prerelease/generate_prerelease_script.py
@@ -23,6 +23,7 @@ from ros_buildfarm.argument import add_argument_os_name
 from ros_buildfarm.argument import add_argument_output_dir
 from ros_buildfarm.argument import add_argument_rosdistro_name
 from ros_buildfarm.config import get_index as get_config_index
+from ros_buildfarm.config import get_release_build_files
 from ros_buildfarm.config import get_source_build_files
 from ros_buildfarm.devel_job import configure_devel_job
 from ros_buildfarm.prerelease import add_overlay_arguments
@@ -78,6 +79,22 @@ def main(argv=sys.argv[1:]):
     build_file = build_files[args.source_build_name]
 
     print('Fetching rosdistro cache...')
+    # Targets defined by source build file are subset of targets
+    # defined by release build files. To increase the number of supported
+    # pre-release targets, we combine all targets defined by all release
+    # build files and use that when configuring the devel job.
+    release_build_files = get_release_build_files(config, args.rosdistro_name)
+    release_targets_combined = {}
+    if release_build_files:
+        release_targets_combined[args.os_name] = {}
+        for build_name, rel_obj in release_build_files.items():
+            if args.os_name not in rel_obj.targets:
+                continue
+            for dist_name, targets in rel_obj.targets[args.os_name].items():
+                if dist_name not in release_targets_combined[args.os_name]:
+                    release_targets_combined[args.os_name][dist_name] = {}
+                release_targets_combined[args.os_name][dist_name].update(targets)
+
     index = get_index(config.rosdistro_index_url)
     dist_cache = get_distribution_cache(index, args.rosdistro_name)
     dist_file = dist_cache.distribution_file
@@ -154,7 +171,8 @@ def main(argv=sys.argv[1:]):
         config=config, build_file=build_file,
         index=index, dist_file=dist_file, dist_cache=dist_cache,
         jenkins=False, views=False,
-        source_repository=source_repository)
+        source_repository=source_repository,
+        build_targets=release_targets_combined)
 
     templates.template_hooks = None
 

